### ISSUE TITLE
Compiler build errors when boot/ocamlc and ./ocamlc produce different cmis

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#9291]
I am creating this issue to track and discuss a build failure that
occurs in the current trunk
when --enable-flambda --disable-flat-float-array are both set. I am
not sure what is the best way forward and hoping for a clarifying
discussion.

### TL/DR: Questions

- Should we explicitly request that boot ocamlc and repo ocamlc
  produce the same .cmi files?

- If not, we have a dependency problem to fix. What do you think would
  be a good approach?

- In particular, would it be reasonable to stop using boot/ocamlc to
  build anything else that the ./ocamlc binary, and build the rest
  with ./ocamlc instead?

Naming:
- `boot/ocamlc`: boot compiler
- `./ocamlc`: repo compiler

### Context

The build failure only started after #2188 was merged, but my current
understanding is that it comes from a defect in the Makefile
dependencies that has presumably been there forever. We only hit
a build failure now because #2188 is such that,
when --disable-flat-float-array is set, the boot compiler
(compiled with --enable-flat-float-array) and the repo compiler
produce different .cmi files for modules of the standard library.

### Problem

The current build process performs the following steps:
1. build the OCaml runtime and bytecode interpreter (ocamlrun)
2. use ocamlrun to build the standard library, using the boot compiler;
  copy stdlib artefacts into boot/ as the "boot stdlib"
3. use the boot compiler and the boot stdlib to build the repo compiler
4. use the repo compiler to build the stdlib again (the "repo stdlib")
5. use the boot compiler and boot stdlib to build the bytecode world
6. use the repo ocamlopt and the repo stdlib to build the native world

The dependency issues is as follows: at the beginning of step (6),
some compiler files that were compiled against the boot stdlib in step
(3) are linked with the repo stdlib built in step (4). This is correct
whenever the two stdlib have the same digests, which apparently was
always the case previously. When --disable-flat-float-array is set,
the bootstrap compiler and the repo compiler produce different digests
since #2188, which produces different type-declaration information
on --disable-flat-float-array.

Note: the reason why the stdlib is rebuilt (step 4) is that
  stdlib/Makefile creates a dependency of stdlib artefacts over the
  compiler currently used to build them (see COMPILER_DEPS in
  stdlib/Makefile). In particular, when we request the "rest of the
  build" after step (3), the repo compiler is newer than the stdlib,
  so this dependency causes a rebuild.

  We cannot fix the present problem with the same trick (by marking
  a dependency of compiler files on the stdlib), because doing so
  would create a dependency cycle: ocamlc depends on compiler files
  that would depend on the stdlib which depends on ocamlc, triggering
  and infinite rebuild whenever ocamlc changes.

### No clear best path to a solution

I considered general several approaches to fix this, including the
following (many of them could turn out to be broken):

1. create a file indicating which phase of the compiler we are in
   (1-2-3-5 are "phase 1", 4-6 are "phase 2"), and having a dependency
   of compiler modules on this phase-file

2. remove the repo-compiler artifacts at the end of (5), to force
   a rebuild (removing artifacts during build sounds like a bad idea)

3. implement machinery to produce build artifacts in "build
   directories" instead of source directories, and use a separate build
   directory (boot/phase1?) for steps 1-2-3-5

These all look like a fair amount of work, so I thought that
discussing the general problem first would be a good first move.

Note: currently, boot/ocamlc remains used to build all core bytecode
  targets, typically ocamlopt (in particular, note that ocaml{c,opt}
  and ocaml{c,opt}.opt are built using different compiler eras, the
  bytecode compilers are build using boot/ocamlc and the native
  compilers with source ocamlopt).

  If we wanted to get out of this general class of issues, I believe
  that we should ensure that boot/ocamlc is only ever used to build
  the ./ocamlc binary, and never afterwards. In particular, ocamlc
  should be used for all followup operations. With that guarantee, we
  could clean the non-boot/ build artifacts after ./ocamlc is
  produced, and not worry about this problem.

### Another way out?

There may be another way out, which is to require that the boot ocamlc
and repo ocamlc only ever produce the same .cmi files. If we had this
property (it generally holds in typical settings), we wouldn't need to
fix our dependency tracking.

For the concrete problem of #2188, we could restore this property by
adding a compile-time flag for {,no-}flat-float-array, for which the
configure-time flag would only select the default. Then we could run
the bootstrap compiler with the configured setting, and thus ensure
that boot/ocamlc and ./ocamlc produce exactly the same .cmi files.

**Comment by gasche** [comment]
cc: @xavierleroy has a good picture of the design choices, and @shindere will want to be in the loop.

**Comment by dra27** [comment]
I have been hacking (in fact, I am literally just now hacking) native cross-compilation into the compiler - the patch-set I'm working on at the moment does start to introduce "out-of-tree" (or at least, "in a different directory") builds to parts of the compiler, because you need all C stubs compiled for two architectures (and potentially three different ocamlrun binaries), and two sets of native code artefacts. So your option 3 is probably where the build is going to end up at some point soon anyway.

**Comment by gasche** [comment]
Yes, I also discussed this out-of-tree builds with @Octachron in the context of code-coverage-by-tests measurements. I agree that it's a desirable state to tend towards, but it is not clear to me that it is the right "fix to develop very soon" for the build-breakage problem we have in trunk today.

**Comment by dra27** [comment]
I did some experiments today - it is reasonably non-invasive to add detection for this situation (using a very similar trick to what `runtime/Makefile` does with `primitives` and `primitives.new` and adding a highly cut-down version of `ocamlobjinfo` to `ocamlc` so you can get the MD5 of .cmi files before compiler libs is built.)

With that, it's also easy to get `boot/ocamlcommon.cma` and so forth (with symlinking in the case where no changes are required), but alas it's not enough. The change in behaviour affects `.cmi` files from the compilerlibs themselves as well. Maintaining a separate copy of all the `.cmi` files in `/boot` feels like a nightmare in terms of a quick fix.

Your description above about the present use of the boot compiler I think is not quite right. The boot compiler is used to build the entirety of the core system, including ocamlopt (the bytecode version of ocamlopt): so it builds, `ocamlc`, `ocaml`, everything in `tools/` except `caml-tex`, `ocamllex` and `ocamlopt` but not otherlibs, the debugger, ocamltest or ocamldoc (all of which are built with the repo stdlib and the repo ocamlc).

**Comment by gasche** [comment]
>  The boot compiler is used to build the entirety of the core system

Indeed, I edited my original message: at step (5), the boot compiler is still used, but with the repo stdlib. (This mismatch seems fairly strange to me.)

**Comment by dra27** [comment]
The bytecode compilation I think always proceeds with the boot stdlib - the only time `-I stdlib` is used in the toplevel is to do native compilation - which is still sound inasmuch as it means you know all the bytecode _tools_ were built with the boot compiler and stdlib.

I'm leaning towards a generalised version of your workaround - there should be an option to feed in alternate contents of the `Config` module to `ocamlc` which I think would fix this. I was also toying with it for cross-compilation, as it would save building multiple ocamlmkconfig and ocamlc images which differ only in terms of configuration... the added use case here is making it seem more compelling (it's also a quickish fix...)

**Comment by gasche** [comment]
The problem is that ocamlc and ocamlopt share some compilation units. For those, building ocamlopt will use the repo stdlib to build the `.cmx` file, while the `.cmi` file was compiled by ocamlc against the boot stdlib.

Here is a lightly edited build log fragment from https://github.com/ocaml/ocaml/pull/2188#issuecomment-581964930
```
$ grep -E "( map.ml|/misc.ml)" log
../boot/ocamlrun ../boot/ocamlc -use-prims ../runtime/primitives [flags...] -o stdlib__map.cmi -c map.mli
../boot/ocamlrun ../boot/ocamlc -use-prims ../runtime/primitives [flags...] -o stdlib__map.cmo -c map.ml
./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives [flags...] -I {utils,parsing,typing...} -c utils/misc.mli
./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives [flags...] -I {utils,parsing,typing...} -c utils/misc.ml
../boot/ocamlrun ../ocamlc [flags...] -o stdlib__map.cmi -c map.mli
../boot/ocamlrun ../ocamlc [flags...] -o stdlib__map.cmo -c map.ml -o stdlib__map.cmx -c map.ml
./boot/ocamlrun ./ocamlopt -g -nostdlib -I stdlib -I otherlibs/dynlink [flags..] -I {utils,parsing,typing...} -function-sections -c utils/misc.ml
```

As you can see, ocamlopt builds utils/misc.ml against the repo stdlib, but utils/misc.mli was built by the boot ocamlc against the boot stdlib.

**Comment by gasche** [comment]
After a short discussion with @Octachron and @damiendoligez, we are considering going the separate-build-dir route. I would like to avoid the horrible Makefile hacks in the style of `dynlink_compilerlibs`; one possible approach would be to rely on the behavior of `-o $(BUILD_DIR)/foo.cm{i,o,x}`, which produces all intermediate files in `$(BUILD_DIR)`.

So I guess that we would have $(BOOT_BUILD_DIR) and $(BUILD_DIR) variables, prefixing all targets, one for artefacts produced by the bootstrap compiler and the other for artifacts produced by repo compilers.

**Comment by shindere** [comment]
Just to mention that as far as I can say `--enable-flambda` plays no
role here. It happens that both `--enable-flambda` and
`--disable-flat-float-array` are tested together by Inria's CI
`other-configs` job, but e.g. on my machine, or on the one of
@damiendoligez, the problem occurs even when flambda is not enabled.

**Comment by shindere** [comment]
Gabriel Scherer (2020/02/11 10:37 +0000):
> After a short discussion with @Octachron and @damiendoligez, we are
> considering going the separate-build-dir route. I would like to avoid
> the horrible Makefile hacks in the style of `dynlink_compilerlibs`;
> one possible approach would be to rely on the behavior of `-o
> $(BUILD_DIR)/foo.cm{i,o,x}`, which produces all intermediate files in
> `$(BUILD_DIR)`.

@gaschde: @damiendoligez and @Octachron may have told you that this
build out of source tree is part of the work we proposed I'd do during
the coming year. That would be a prerequisite for cross-compilation, as
@dra27 has explained.

Another part of the work we have proposed consists in moving from the
calling make recursively to including sub-makefiles, so that make gets a
better view of dependencies. I think we will need to discuss to make
sure we don't do work twice.

At the moment, it feels to me if it would be possible to make the two
compilers produce the same `.cmi` files it would be a nice way to go for
the time being.

**Comment by shindere** [comment]
So I'll work on this ASAP (discussed with @gasche and @damiendoligez).

**Comment by dra27** [comment]
In between things, I've been hacking a prototype, which suggests the diff is not as big as might be being feared - in fact, so far it's easier to debug with `-I boot` instead of 15 `-I` directories...!

**Comment by gasche** [comment]
Is the code of your prototype pushed to a WIP-branch somewhere?

My current understanding of the "shortest clean solution" would be to:
- have separate build-artifact directories for boot-compiler artifacts and for repo-compiler artifacts. (I considered keeping the repo-compiler artifacts in the source tree, which is pleasingly the same behavior as today but runs into more risk of including them by mistake in a boot-compiler build due to the fact the implicit `-I .`.)
- do this only for ocaml{c,opt} compilation units; instead of doing the same work in otherlibs and all other subsystems, we could just move them to always use the repo-compiler

I'm not sure about:
- whether the build repositories should be a flat directory for all compilation units, or should mirror the directory structure of the source directory (your point on `-I` suggests that you took the first route)
- how to deal with `.ml` artefacts (produced by ocamllex or yacc/menhir); it's easier to keep them in the source tree but more consistent to have them in the build tree

I also hope that as a result we will need to bootstrap less (because the boot-compiler and repo-compiler cmis don't have to be identical), and I am not sure how to test this.

**Comment by dra27** [comment]
I’ll push it and sync up with @shindere in an hour or so

**Comment by xavierleroy** [comment]
I'm catching up with this discussion.

The "build in multiple directories" approach seems quite heavy to me.  It's basically a bootstrap.  Doing that for every build is wasteful.  Plus, our build system is complicated enough already.

My preferred approach would be to make sure that boot/ocamlc and ./ocamlc produce compatible .cmi files.  (Unless in transitory states where a bootstrap is required, of course.)  I see two ways to preserve this property in the context of #2188:

- Add a command-line option to set the "flat float array" mode of ocamlc, overriding the default stored in boot/ocamlc. Just like the `-use-prims` flags overrides the default list of C primitives stored in boot/ocamlc.

- Revisit the implementation of #2188 so that it does not store data representation information in type constructors, but recomputes it when needed from the definitions of the type constructors.  The "don't store but derive from the types when needed" approach is the approach we've followed so far, e.g. in https://github.com/ocaml/ocaml/blob/trunk/typing/typeopt.ml .

**Comment by xavierleroy** [comment]
Paging Dr. @Octachron, release manager: I don't think we can release 4.11 with a broken build in disable-flat-float-array mode, so either this is fixed quickly or #2188 will have to be reverted from the 4.11 branch.

**Comment by gasche** [comment]
One thing that I like with "multiple build directories" is that it would help with many other things, including:
1. cross-compilation (to avoid mixing object files written for the host and target systems)
2. nice interaction with Dune, or some other approach to also build the codebase with a released compiler for better Merlin support
3. the ability to bootstrap the compiler from another OCaml implementation that uses a different serialization format

I agree that the main problem with it is build system complexity, but due to (1) this is something that may come to our place anyway. Build complexity is the reason I have not worked on it myself yet. In terms of wastage, yes, that essentially means building `ocamlc` twice. (But note that due to (2) we can hope that our development builds would build the first stage using the .opt binaries in our PATH, which could be pretty fast).

In any case: you are very right to point out that we need a solution for 4.11. One short-term solution for #2188 that I am considering is to compute the separability information in the same way when `-no-flat-float-array` is set, to get identical `.cmi`, and then ignore the information (instead of using it to reject "dangerous" definitions, as they are not dangerous on this runtime). This is unpleasant as the no-flat-float systems should not require computing the information at all, but it may still be a reasonable workaround for the time being. (It also makes it possible to provide a warning if we accept a definition that would be rejected in `-flat-float-array` mode, if users ask for this at some point.)

Regarding your two proposals:

- The problem with a `-flat-float-array` mode is that the configuration affects the runtime and the code generation strategy. What we would need here (to produce valid binaries and avoid `.cmx` issues, etc.) is "do the static checks as in flat-float-array but otherwise assume no-flat-float-array", which is  would only be justified by unpleasant internal reasons.

- Unlike the typeopt-computed properties, the separability information is computed by a fixpoint, essentially like variance, so it seems wasteful (and not conceptually quite right) to recompute it each time.

**Comment by xavierleroy** [comment]
> compute the separability information in the same way when -no-flat-float-array is set, to get identical .cmi, and then ignore the information (instead of using it to reject "dangerous" definitions, as they are not dangerous on this runtime).

Sounds good to me.  Do you think you could do this in time for the next alpha/beta release of 4.11?

Re: multiple build directories, I'm not convinced yet, and the design needs to be articulated precisely, so let's keep that for later.

Re: on-demand computation of fixpoints, that could be expensive in some cases, but I have the impression that the solution is trivial most of the time.  Also, there are [clever libraries for lazy fixpoint computations](https://gitlab.inria.fr/fpottier/fix).

Re: the `-flat-float-array` option, in my mind it was one of those undocuments, ocamlc-specific options whose only purpose is to smooth out the bootstrap, like `-use-prims` today.  In particular, it would not be supported by `ocamlopt` and could not affect native-code generation.

There's also a more general alternative: instead of "baking in" configuration options in the Config module which is then statically linked in the bootstrap compiler, put the configuration options in a text file, say `ocaml.ini`, located at a fixed relative position from `ocamlc` and `ocamlopt`, and have these executables read the configuration parameters from this text file.  This way, the proper `ocaml.ini` is generated by configure or by the build system, and the archived `boot/ocamlc` is made to use this `ocaml.ini`, just like the generated `./ocamlc`.

If you're curious about the latter approach: it's the one we use in CompCert to support binary distributions where the installed can update the `compcert.ini` file appropriately.  This is what it looks like: https://github.com/AbsInt/CompCert/blob/master/driver/Configuration.ml .

**Comment by gasche** [comment]
Yes, the "relocatable configuration" is something we discussed in the past and, I also agree, would be good and solve other problems as well. (And it's simpler than hacking the build system.) (For example, it may also help with cross-compilation, by making it easier to use the same compiler to generate code for the host system when we want, and for the target system when we want.)

**Comment by nojb** [comment]
Oops sorry closed by mistake.

**Comment by gasche** [comment]
Re. "relocatable configuration" (reading the compiler configuration from a file, instead of a configure-generated static module): this looks like something that @shindere might be interested in looking at.

**Comment by gasche** [comment]
Note: I propose a fix for the #2188-specific source of `.cmi` difference in #9592.
(I believe this specific issue is blocking for the 4.11 release.)

**Comment by shindere** [comment]
Gabriel Scherer (2020/05/22 00:53 -0700):
> Re. "relocatable configuration" (reading the compiler configuration
> from a file, instead of a configure-generated static module): this
> looks like something that @shindere might be interested in looking at.

Why not, in the long run.

**Comment by Octachron** [comment]
I removed the 4.11 milestone since this specific problem has been surgically removed for now.

**Comment by gasche** [comment]
Here is my take-away from the state of the discussion:

- (1) I would like to be able to support situation where different compiler configurations produce different `.cmi` (in the same format), but there is no consensus there, Xavier seems happy with having the same-cmi-files constraint.
- (1 bis) In any case our build system is too painful for me to deal with right now, so any change in this direction will have to wait for a braver developer or a brighter build-system future.
- (2) The recurring suggestion to read our configuration at runtime instead of baking it in the sources at configure-time keeps making sense and would improve on many other situations. (In this case I'm not sure it is the best approach, but it's useful in other cases). I think that @nojb looked at it not-so-long ago, maybe in a PR that we shamefully failed to provide feedback on?

**Comment by shindere** [comment]
Gabriel Scherer (2021/06/28 05:25 -0700):
> Here is my take-away from the state of the discussion:
>
> - (1) I would like to be able to support situation where different compiler configurations produce different `.cmi` (in the same format), but there is no consensus there, Xavier seems happy with having the same-cmi-files constraint.
> - (1 bis) In any case our build system is too painful for me to deal
> with right now, so any change in this direction will have to wait
> for a braver developer or a brighter build-system future.

I hope to be able to dealwith that, withthe help of David also.

**Comment by shindere** [comment]
github-actions[bot] (2021/06/27 21:25 -0700):
> This issue has been open one year with no activity.  Consequently, it
> is being marked with the "stale" label.  What this means is that the
> issue will be automatically closed in 30 days unless more comments are
> added or the "stale" label is removed.  Comments that provide new
> information on the issue are especially welcome: is it still
> reproducible? did it appear in other contexts? how critical is it?
> etc.

As a consequence to the previous coment, I'd suggest to keep this open,
but if it bothers people I don't see a problem in closing it.

Suggestion: could someone who feels more comfortable than me with
GitHub's interface perhaps assign this to me explicitly? It's not a
commitment that I'll act on it in a predictable future, but I expect
this to be solved as side effects of different clenaup and refactoring
tasks on the build system.

**Comment by dra27** [comment]
> Suggestion: could someone who feels more comfortable than me with GitHub's interface perhaps assign this to me explicitly? It's not a commitment that I'll act on it in a predictable future, but I expect this to be solved as side effects of different clenaup and refactoring tasks on the build system.

It's assigned to you! (and is still open)

**Comment by shindere** [comment]
David Allsopp (2021/08/09 07:40 -0700):
> Assigned #9291 to @shindere.

thanks a lot, David!

**Comment by shindere** [comment]
David Allsopp (2021/08/09 07:41 -0700):
> > Suggestion: could someone who feels more comfortable than me with GitHub's interface perhaps assign this to me explicitly? It's not a commitment that I'll act on it in a predictable future, but I expect this to be solved as side effects of different clenaup and refactoring tasks on the build system.
>
> It's assigned to you! (and is still open)

I think it's the right way to go. thanks again.

**Comment by shindere** [comment]
Still an issue but we are progressing slowly towards its clean
resolution with the work on the build system.
